File: src/core/src/gen-constants.c
#include <gmp.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "constants.h"

bool isCoprime(mpz_t a, mpz_t b) {
  mpz_t gcd;
  mpz_init(gcd);
  mpz_gcd(gcd, a, b);

  bool coprime = mpz_cmp_ui(gcd, 1) == 0;

  mpz_clear(gcd);

  return coprime;
}

bool findModularInverse(mpz_t result, mpz_t num, mpz_t mod) {
  mpz_t gcd, x, y;
  mpz_inits(gcd, x, y, NULL);

  mpz_gcdext(gcd, x, y, num, mod);

  bool inverseExists = mpz_cmp_ui(gcd, 1) == 0;

  if (inverseExists) {
    mpz_mod(result, y, mod);
    if (mpz_sgn(result) < 0) {
      mpz_add(result, result, mod);
    }
  }

  mpz_clears(gcd, x, y, NULL);

  return inverseExists;
}

int main() {
  srand(time(NULL));

  char nString[BOOK_LENGTH];

  int n;
  for (n = 0; n < BOOK_LENGTH; n++) {
    sprintf(&nString[n], "%c", 's');
  }

  mpz_t N;
  mpz_init(N);
  mpz_set_str(N, nString, ALPHA_LENGTH);

  char randomStartNum[BOOK_LENGTH];
  randomStartNum[0] = 's';

  for (n = 1; n < BOOK_LENGTH; n++) {
    char randomChar = BASE29_ALPHA[rand() % ALPHA_LENGTH];
    sprintf(&randomStartNum[n], "%c", randomChar);
  }

  mpz_t C;
  mpz_init(C);
  mpz_set_str(C, randomStartNum, ALPHA_LENGTH);

  while (1) {
    bool foundCoprime = isCoprime(N, C);

    if (foundCoprime) {
      mpz_t modInv;
      mpz_init(modInv);
      bool foundModInv = findModularInverse(modInv, N, C);

      if (foundModInv) {
        char nString[BOOK_LENGTH + 1];
        char cString[BOOK_LENGTH + 1];
        char iString[BOOK_LENGTH + 1];

        mpz_get_str(nString, 29, N);
        mpz_get_str(cString, 29, C);
        mpz_get_str(iString, 29, modInv);

        printf("%s\n%s\n%s", nString, cString, iString);

        mpz_clear(modInv);

        break;
      } else {
        printf("modular inverse does not exist\n");
        mpz_clear(modInv);
      }
      
    }

    mpz_sub_ui(C, C, 1);
  }

  mpz_clear(N);
  mpz_clear(C);
  return 0;
}

File: src/utils/gen-constants.ts
import fs from "fs";
import { init as gmp_init, GMPFunctions, mpz_ptr } from "gmp-wasm";
import { ALPHA, BOOK_LENGTH, BASE32_ALPHA, BASE32_LAST } from "../constants";

async function isCoprime(binding: GMPFunctions, a: mpz_ptr, b: mpz_ptr) {
  const gcd = binding.mpz_t();
  binding.mpz_init(gcd);
  binding.mpz_gcd(gcd, a, b);

  return binding.mpz_cmp_ui(gcd, 1) === 0;
}

async function findModularInverse(
  binding: GMPFunctions,
  num: mpz_ptr,
  mod: mpz_ptr
) {
  const gcd = binding.mpz_t();
  binding.mpz_init(gcd);

  const x = binding.mpz_t();
  binding.mpz_init(x);

  const y = binding.mpz_t();
  binding.mpz_init(y);

  binding.mpz_gcdext(gcd, x, y, num, mod);

  const inverseExists = binding.mpz_cmp_ui(gcd, 1) === 0;

  const result = binding.mpz_t();
  binding.mpz_init(result);

  if (inverseExists) {
    binding.mpz_mod(result, y, mod);
    if (binding.mpz_sgn(result) < 0) {
      binding.mpz_add(result, result, mod);
    }
  }

  return { inverseExists, result };
}

(async () => {
  const { binding } = await gmp_init();

  const N = binding.mpz_t();
  binding.mpz_init(N);
  binding.mpz_set_string(
    N,
    new Array(BOOK_LENGTH).fill(BASE32_LAST).join(""),
    ALPHA.length
  );

  let randomStartNum = BASE32_LAST;
  for (let i = 1; i < BOOK_LENGTH; i++) {
    randomStartNum +=
      BASE32_ALPHA[Math.floor(Math.random() * BASE32_ALPHA.length)];
  }

  const C = binding.mpz_t();
  binding.mpz_init(C);
  binding.mpz_set_string(C, randomStartNum, ALPHA.length);

  while (true) {
    const foundCoprime = await isCoprime(binding, N, C);

    if (foundCoprime) {
      const { result, inverseExists } = await findModularInverse(binding, N, C);

      if (inverseExists) {
        const nString = binding.mpz_to_string(N, ALPHA.length);
        const cString = binding.mpz_to_string(C, ALPHA.length);
        const iString = binding.mpz_to_string(result, ALPHA.length);

        fs.writeFileSync("numbers", `${nString}\n${cString}\n${iString}`);

        break;
      } else {
        console.log("modular inverse does not exist");
        break;
      }
    }

    binding.mpz_sub_ui(C, C, 1);
  }

  await binding.reset();
})();
